smolyay/surrogate.py

```diff
@@ -1,546 +1,828 @@
+        # reset constant
+        self._integration_constant = 0
+        if not self._terms_constructed_cache:
+            self._create_terms()
+            self._terms_constructed_cache = True
+        # get points
+        if isinstance(
+            X,
+            (
+                smolyay.samples.UnidimensionalPointSet,
+                smolyay.samples.MultidimensionalPointSet,
+            ),
+        ):
+            X = X.points
+
+        # validate data inputs
+        X = numpy.array(X, ndmin=2)
+        if X.shape[1] != self.num_dimensions:
+            raise IndexError("Must be 2D array with shape (n_samples, n_features)")
+        y = numpy.array(y, ndmin=1)
+        if y.shape != (X.shape[0],) and y.shape != (X.shape[0], 1):
+            print(y.shape)
+            raise IndexError("Must be 2D array with shape (n_samples,)")
+
+        oob = any(
+            numpy.any(X[:, i] < self.domain[i][0])
+            or numpy.any(X[:, i] > self.domain[i][1])
+            for i in range(self.num_dimensions)
+        )
+        if oob:
+            raise ValueError("X must lie in domain of surrogate")
+        # create basis matrix
+        num_basis_max = numpy.max([len(p) for p in self._basis_sets])
+        if any(
+            any(bf._is_complex for bf in basis_set) for basis_set in self.basis_sets
+        ):
+            lookup_table = numpy.zeros(
+                (self.num_dimensions, num_basis_max, len(X)), dtype="complex_"
+            )
+            basis_matrix = numpy.zeros(
+                (len(X), len(self._index_combinations)), dtype="complex_"
+            )
+        else:
+            lookup_table = numpy.zeros((self.num_dimensions, num_basis_max, len(X)))
+            basis_matrix = numpy.zeros((len(X), len(self._index_combinations)))
+        # solve for the inputs at all the basis functions
+        for dim in range(self.num_dimensions):
+            for i, basis_fun in enumerate(self.basis_sets[dim]):
+                new_X = basis_fun.domain[0] + (
+                    basis_fun.domain[1] - basis_fun.domain[0]
+                ) * (
+                    (X[:, dim] - self.domain[dim, 0])
+                    / (self.domain[dim, 1] - self.domain[dim, 0])
+                )
+                numpy.clip(new_X, basis_fun.domain[0], basis_fun.domain[1], out=new_X)
+                lookup_table[dim, i, :] = basis_fun(new_X)
+
+        # use lookup table to solve for each term
+        for term, ic in enumerate(self._index_combinations):
+            basis_matrix[:, term] = numpy.prod(
+                [lookup_table[dim, ic[dim], :] for dim in range(len(ic))], axis=0
+            )
+
+        # solve for coefficients
+        if self.regularization is None or numpy.any(numpy.iscomplex(basis_matrix)):
+            if basis_matrix.shape[0] == basis_matrix.shape[1]:
+                try:
+                    self._coefficients = numpy.linalg.solve(basis_matrix, y)
+                except:
```

Comment:
You should check for the specific exception that is raised when the solve fails, not all exceptions.